### ISSUE TITLE
verkko trio mode failed

### ISSUE THREAD

**Issue by LHG-GG** [marbl/verkko#194]
Hi @skoren,

I am running an assembly (human sample) in a trio mode with latest version v1.4.1
`TMPDIR='./tmp' time verkko -d trio_results --hifi I002_Duplex_10k_reads.fastq.gz --nano I002C_T2T_ULONT_100kb.fastq.gz --hap-kmers data/maternal_compress.k30.hapmer.meryl data/paternal_compress.k30.hapmer.meryl trio --keep-intermediate --local-cpus 128 --local-memory 700`

I encountered the same error as https://github.com/marbl/verkko/issues/186, https://github.com/marbl/verkko/issues/185. So, I restarted the assembly using v1.4.0 on the same folder. This time assembly completed but results are not as expected.  

The `assembly.fasta` is around expected size (6.3 gb) but both `haplotype fastas` are ~500MB and majority of the sequences are in `assembly.unassigned.fasta` (5.2gb). Not sure what has gone wrong. Appreciate your help in this regards. 

Thank you

**Comment by skoren** [comment]
It's likely either the haplotypes are too similar to have no markers, there is an issue with the parents, or with the databases you're using. If you can share the final graph and its coloring, it should give more info on what is the issue. Both are in the 6-rukki folder, named `unitig*noseq.gfa` and `unitig*csv`

**Comment by LHG-GG** [comment]
Hi,

_>>If you can share the final graph and its coloring, it should give more info on what is the issue._
I have attached

_>> Also run meryl print data/paternal_compress.k30.hapmer.meryl |head to confirm the database is in compressed space_
![image](https://github.com/marbl/verkko/assets/140685334/b2db2244-b7ae-4e91-9838-e02418b381a1)

_>>It's likely either the haplotypes are too similar to have no markers, there is an issue with the parents, or with the databases you're using_
I run yak triobin on assembly.fasta; it looks like contigs have parental specific k-mer. Similar graph is generated with merqury as well (used uncompressed parental db)
![VTCF_blob](https://github.com/marbl/verkko/assets/140685334/44220653-2845-4838-bdd0-70bc475421dc)
[unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.colors.csv](https://github.com/marbl/verkko/files/12658408/unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.colors.csv)

**Comment by skoren** [comment]
The colors.csv definitely supports that something is wrong with the databases. Almost none of the nodes have markers and those that do have very few despite being megabases in size. Here's an example:
<img width="939" alt="Screen Shot 2023-09-19 at 10 24 57 AM" src="https://github.com/marbl/verkko/assets/923478/84a94a45-683d-4432-b09a-9625a8b510fa">

Can you compare the marker counts for one node in assembly.fasta vs what you have from the compressed databases? For example, utig4-11197 and utig4-11198, look those up in `6-*/*scfmap` and check the counts of those final contigs in the assembly.fasta. If they are significantly different, that would confirm something went wrong with the compressed DBs. Are there any errors in `6-rukki/rukki.err` reported?

**Comment by DustinSokolowski** [comment]
Hello!

I think that I am coming into the a very similar issue (hap1 is 120Mb, hap2 is the correct size (2.5Gb), unplaced is 2.1Gb, and I think the root of it (on my end) could have to do with one of the two issues:

1 - I do not have child data, only maternal and paternal 
2 - I have 10X linked read data, and compressing and calculating the union-sum for R1 and R2 separately could be causing downstream issues with the hapmer file.

The script I'm running (see below) runs fine in it's individual steps. I ran the final union-sum line `$MERYL/meryl union-sum k=30 threads=16 memory=350 maternal.compress.k30.meryl paternal.compress.k30.meryl output child.compress.k30.meryl` because the `hapmers.sh` script does not work unless the `child.compress.k30.meryl` file is present.

Is there any chance you can provide some suggestions on how to deal with these potential issues?

Thanks,
Dustin

```
#zcat $input | awk '{if (NR%2==1) {print $1} else {print substr($1,24)}}' | meryl k=$k count output $output -

mkdir -p /scratch2/groups/simpsonlab/dsokolowski/fq1

# Materal R1 meryl dir
#  zcat $matFQ/*_R1_001.fastq.gz | awk '{if (NR%2==1) {print $1} else {print substr($1,24)}}' > /scratch2/groups/simpsonlab/dsokolowski/fq1/maternal_R1_trimmed.fastq

#  $MERYL/meryl count compress k=30 threads=16 memory=400 /scratch2/groups/simpsonlab/dsokolowski/maternal_R1_trimmed.fastq output maternal.k30.R1.meryl
#  $MERYL/meryl count compress k=30 threads=16 memory=400 $matFQ/*_R2_001.fastq.gz output maternal.k30.R2.meryl

# Maternal R2 meryl dir
#  zcat $patFQ/*_R1_001.fastq.gz | awk '{if (NR%2==1) {print $1} else {print substr($1,24)}}' > /scratch2/groups/simpsonlab/dsokolowski/fq1/paternal_R1_trimmed.fastq

# $MERYL/meryl count compress k=30 threads=16 memory=400 /scratch2/groups/simpsonlab/dsokolowski/paternal_R1_trimmed.fastq output paternal.k30.R1.meryl
# $MERYL/meryl count compress k=30 threads=16 memory=400 $patFQ/*_R2_001.fastq.gz output paternal.k30.R2.meryl

# Combine R1 and R2
# $MERYL/meryl union-sum k=30 threads=16 memory=350 maternal.k30.R1.meryl maternal.k30.R2.meryl output maternal.compress.k30.meryl

# $MERYL/meryl union-sum k=30 threads=16 memory=350 paternal.k30.R1.meryl paternal.k30.R2.meryl output paternal.compress.k30.meryl

$MERYL/meryl union-sum k=30 threads=16 memory=350 maternal.compress.k30.meryl paternal.compress.k30.meryl output child.compress.k30.meryl

$MERQURY/trio/hapmers.sh \
  maternal_compress.k30.meryl \
  paternal_compress.k30.meryl \
     child_compress.k30.meryl
```

**Comment by skoren** [comment]
The child reads are optional to hapmers.sh, you can just leave them out or use k-mers from the child's HiFi data instead. The fact that one of the haplotypes is OK implies there may be an issue with one of the parental data types. hapmers.sh will produce some plots to visualize the parent-specific k-mer distributions, could you post those here? Your graph (`assembly.noseq*.gfa`) and the colors file (`assembly.colors.csv`) would also help to visualize where the markers are being found in the assembly.

**Comment by DustinSokolowski** [comment]
Hey Sergey,

Thank you so much for the quick reply. Please find the files attached. Looking at the plots, I'm wondering if the union-sum is inflating the number of child-only hapmers? Since we used the union-sum would everything end up being shared or child-only, thereby screwing things up?

Strangely, the hapmer.sh script actually throws an error if I don't have the child kmer data as an option. Specifically it has a hard time finding the meryl/merqury directories and skips steps, which makes me think that it's some sort of flow control thing. Regardless, I'll get k-mers from the child's HiFi data as you suggested and rerun the pipeline & update.

[verkko_trio_debug.zip](https://github.com/marbl/verkko/files/14982296/verkko_trio_debug.zip)

![inherited_hapmers fl](https://github.com/marbl/verkko/assets/56414662/6ada932c-06eb-4f06-bac1-c6b168ce9a98)
![inherited_hapmers ln](https://github.com/marbl/verkko/assets/56414662/f946ae3f-d11f-460b-a996-542123086d15)
![inherited_hapmers st](https://github.com/marbl/verkko/assets/56414662/29090b16-9289-44f0-9c4f-668d62cf5876)

**Comment by DustinSokolowski** [comment]
Good evening,
Quick update if it helps, I counted k-mers with the child's hifi data and reran the hapmer script. The parent-specific distributions look more reasonable, albeit quite shared. This isn't a huge surprise as my species (naked mole-rats) were wild derived ~30 years ago and have been inbred since then. 
![inherited_hapmers st](https://github.com/marbl/verkko/assets/56414662/c22b5065-7482-4b05-ab8d-25cb59a41260)

Also when trying to re-run the hapmer script without the child data at all made it to "maternal.compress.k30_and_paternal.compress.k30.meryl" before ending.

Unless I'm getting mixed up, I think that the check 
```
if [[ -z $child_meryl ]]; then
	echo "No child.meryl given. Done!"
	exit 0
fi
```
happening at line 64-67 (which is before the "inherited.gt2.meryl" files are created is the reason the hapmer files aren't outputted without the child as input.

Best!
Dustin

**Comment by skoren** [comment]
I think there was definitely something wrong with the previous markers, there is almost nothing with maternal markers in the assembly, including in regions where you have a clear haplotype bubble with one side having paternal markers but the other side having no markers. 
<img width="753" alt="Screenshot 2024-04-17 at 3 22 54 PM" src="https://github.com/marbl/verkko/assets/923478/84e3d90d-6aa8-42cf-9339-c6dd6e01005b">
Overall the graph is also very fragmented, is this a HiFi-only assembly? There are relatively short nodes (30-40kb) that are homozygous and would normally be resolved via ONT data.

Hopefully the new markers will address this. You can quickly test this by updating the old marker databases and re-running `6-rukki/rukki.sh` or updating the path to the marker database in the shell script and re-running. You can then load the graph in BandageNG and load the colors file and see if you have both maternal/paternal colors. Another issue is due to the high homozygosity of the genome. There are some very large homozygous nodes (based on graph structure and coverage) that are almost 20 Mbp. Verkko doesn't expect such large nodes to be homozygous in the v2.0 release. We plan to update this in the next release but you could also add `params="$params --max-homozygous-len=25000000"` to the rukki.sh script before you re-run. Once that is done and you confirmed the coloring (or post the resulting color file here), you should be able to erase all files named `assembly.*`, `6-layoutContigs`, and `7-consensus` and re-run verkko and it should make new consensus with the updated phasing results.

**Comment by DustinSokolowski** [comment]
Hey Sergey,

I do think that the new markers helped address this issue, the binning into each haplotype is balanced now. I've attached the colors and graph files as well. I'm going to redo the rukki step with `params="$params --max-homozygous-len=25000000"` and see what that updates as well. As such the attached graph is the fix for the first issue, but does not yet have the params="$params --max-homozygous-len=25000000" change.

<img width="298" alt="Screenshot 2024-04-18 at 11 24 33 AM" src="https://github.com/marbl/verkko/assets/56414662/12272485-0d83-4814-9773-09883a585067">

[verkko_eval.zip](https://github.com/marbl/verkko/files/15027123/verkko_eval.zip)

 The run has one PB-HiFi run (25x) and one ONT R10.4 run (40x). We have two more ONT runs which I have tried adding now that the trio binning issue I was having looks resolved. This being said, it runs into a seg_fault error at the "combineOverlapsConfigure.sh" step when I use all three ONT runs. Do you know which memory flag is needed for that?

**Comment by skoren** [comment]
Yes the markers look improved here though they are still pretty sparse which I guess is a consequence of the relatively homozygous genome and parents. The ONT data wouldn't change anything in combineOverlapsConfigure.sh as it's not used until the 4-processONT step so I don't think this is related since it's only input as --nano and the initial graphs are built from HiFi data. In fact, you can reuse everything from 0-correction to 2-processGraph for the run w/more ONT data.

**Comment by DustinSokolowski** [comment]
Sounds good! I reached out to the individuals who collected the animals to see if they have any details on the colony structure (since naked mole-rat colonies are complex). We know for sure that our samples parents are their actual parents, but I'm not sure if the queen (maternal) and consort (paternal) are littermates, or the somewhat likely possibility that the consort (paternal) is also the queens son, making the paternal and child half-siblings.

That is interesting RE the segfault. I've gotten verkko to reach an output with 8 different inputs (including trio's with one ONT run and non-trio's with three ONT runs) and I haven't come across this issue before. Perhaps I was getting close to the max-mem on the other runs and this got it over the line. Do you know what the flag to up the memory for the "combineOverlapsConfigure" step is @Dmitry-Antipov ? Right now the segfault is coming with ` mem_mb=100205, mem_mib=95563, disk_mb=100205, disk_mib=95563, tmpdir=<TBD>, job_id=1, n_cpus=1, mem_gb=128, time_h=18`

Thanks again!

**Comment by skoren** [comment]
The flags are from `--ovs-run` but I am not sure the memory would cause the segfault, especially since it's already using 128gb. What's the verkko command you're using and what is the log up to the segfault in combineOverlapsConfigure.err?

**Comment by DustinSokolowski** [comment]
Please find the command and log (u
[2024-04-17T104309.617922.snakemake.log](https://github.com/marbl/verkko/files/15028023/2024-04-17T104309.617922.snakemake.log)
p to the segfault) attached. 

```
verkko -d /scratch2/groups/simpsonlab/dsokolowski/PB_trio_allONT --hifi $fqDir/m84134_240306_201715_s1.fastq --nano $fqDir/*_lsk114_run1.fastq --sge --par-run 8 32 72  -
-hap-kmers $trioDir/maternal.compress.k30.hapmer.meryl $trioDir/paternal.co
[2024-04-18T123846.720610.snakemake.log](https://github.com/marbl/verkko/files/15028005/2024-04-18T123846.720610.snakemake.log)
mpress.k30.hapmer.meryl trio
```

Looking at it, it seems like the compute-overlaps command didn't actually finish calculating all of the overlaps, so the issue is probably derived from that.

**Comment by DustinSokolowski** [comment]
Quick udpate here. The error did come from our scheduler cancelling some of the previous "overlap-jobs", making it impossible to compute the overlaps and the rest of the pipeline ran smoothly. Thank you so much for your continued patience and support, I can't wait to see how the genomes look.

Best,
Dustin
